Improvements for Distributed

kfernandez31 opened the pull request:
This branch introduces:
- missing safety checks (kind of a big deal),
- more types of maps for a more ergonomic API. The ones that existed before now have their `_mut` counterparts, which allow for regular mutability, instead of interior mutability. Note, however, that `stop` still requires interior mutability, should you want to alter the service's state.
- the much-awaited functionality of `peering_sharded_service`, through which instances can communicate and delegate work.

Commit:
Add PeeringShardedService

Commit:
Add missing safety checks

Commit:
Add more types of maps, including mut-maps

piodul commented on seastar/src/distributed.rs:
```diff
@@ -303,6 +311,33 @@ impl<S: Service> Distributed<S> {
     }
 
+    fn submit_to_mut<'a, Func, Fut, Ret>(
+        &'a self,
```
This is very unsafe. This allows creating mutable reference from an immutable reference without any checks. If `submit_to_mut` is called twice in a row then it can result in two tasks being created with a `&mut` reference to the same `S`, and this is undefined behavior.

This can only work if this function takes self as `&'a mut self`.

piodul commented on seastar/src/distributed.rs:
```diff
@@ -273,6 +275,8 @@ impl<S: Service> Distributed<S> {
         Ret: Send + 'static,
         I: IntoIterator<Item = u32>,
     {
+        crate::assert_runtime_is_running();
```
The `stop` method is also missing such a check, no?

piodul commented on seastar/src/ffi_utils.rs:
```diff
@@ -66,3 +66,10 @@ pub const fn get_dropper_const<T>(_: &T) -> fn(*const u8) {
+pub struct PtrWrapper(pub *mut u8);
+
+unsafe impl Send for PtrWrapper {}
+unsafe impl Sync for PtrWrapper {}
```
This API is unsafe but allows itself to be used from safe code. That shouldn't be possible - you should add appropriate protections.

For example:

- Only allow creating a `PtrWrapper` via `unsafe fn new(ptr: *mut u8)`
- Only accessing the inner pointer via `unsafe fn as_ptr_mut(&self) -> *mut u8`.

piodul commented on seastar/src/distributed.rs:
```diff
@@ -276,6 +276,33 @@ impl<S: Service> Distributed<S> {
         ffi::stop(self._inner.as_ref().unwrap()).await.unwrap();
     }
 
+    fn submit_to<'a, Func, Fut, Ret>(
```
`make maps create instantly running futures`

How about a test?

piodul commented on seastar/src/distributed.rs:
```diff
@@ -279,12 +291,21 @@ impl<S: Service> Distributed<S> {
                     let local = ffi::local(inner.as_ref().unwrap());
                     let local = unsafe { &*(local as *const S) };
-                    func(local).await
+                    let _ = &distr; // this is to avoid a partial move of the pointer
```
Could you elaborate? I don't understand what a "partial move" means here.

kfernandez31 replied:
Without that line, `rustc` performed a partial move on the tuple-struct's pointer field.

piodul reviewed (changes requested):
The first three commits look OK, but I don't think `map_mut` is going to work like this.

First and foremost, you must ensure exclusive access to the shard-local instances of the service. Unless you ensure this dynamically (e.g. via synchronization, which I'm not sure is the responsibility of `Distributed`), if you take `Distributed` by immutable reference then you risk creating two mutable references to the same shard-local instances, which is undefined behavior. The `_mut` methods need to take the whole `Distributed` instance via mutable reference.

Another thing is the "Peering" part. Due to the problems highlighted above, I don't think `map_mut` and friends should provide access to the container because it would allow creating more references to the same shard-local instances.

Commit:
Make maps create instantly running futures

kfernandez31 commented:
v2 - resolved @piodul 's comments apart from this one:
> Another thing is the "Peering" part. Due to the problems highlighted above, I don't think map_mut and friends should provide access to the container because it would allow creating more references to the same shard-local instances.

I'd like to know whether the security checks I added have addressed your concerns or should I go ahead and rescind the access to the container of mut-maps.

piodul commented on seastar/src/distributed.rs:
```diff
@@ -291,11 +312,20 @@ impl<S: Service> Distributed<S> {
+            let lock = lock.try_read();
+            if lock.is_err() {
+                panic!(
+                    "Data race detected: can't lock instance {} for read-access",
```
The message is misleading - it's not a data race. Data race happens when two or more threads try to access the same data in an unsynchronized way. Here, the lock will always be taken by only one thread - so it's more like you prevent a violation of Rust's aliasing rules. Perhaps you could make the message more similar to what a `RefCell` generates when it detects a rules violation.

Anyway, a more useful error message would also mention possible reasons for why the panic happened so that the developer quickly understands what they did wrong.

piodul commented on seastar/src/distributed.rs:
```diff
@@ -513,6 +733,7 @@ mod tests {
 
     #[seastar::test]
     async fn test_map_all() {
+        //TODO
```
Leftover TODO?

piodul commented on seastar/src/distributed.rs:
```diff
@@ -525,8 +746,29 @@ mod tests {
 
+    #[seastar::test]
+    async fn test_map_all_mut() {
+        //TODO
```
Leftover TODO?

piodul commented on seastar/src/distributed.rs:
```diff
@@ -525,8 +746,29 @@ mod tests {
+
     #[seastar::test]
     async fn test_map_others() {
+        //TODO
```
More unexplained todos...

kfernandez31 replied:
yup, I definetely missed them :p

piodul commented on seastar/src/distributed.rs:
```diff
@@ -304,6 +334,42 @@ impl<S: Service> Distributed<S> {
     }
 
+    fn submit_to_mut<'a, Func, Fut, Ret>(
+        &'a self, // this immutable reference is intentional
```
Why is it intentional? This comment piques reader's curiosity but doesn't satisfy it.

kfernandez31 replied:
True.

piodul commented on seastar/src/distributed.rs:
```diff
@@ -100,15 +103,32 @@ where
 pub struct Distributed<S: Service> {
     _inner: SharedPtr<distributed>,
     _ty: PhantomData<S>,
+    _locks: Vec<Arc<RwLock<()>>>,
```
Seeing a synchronous `RwLock` here looks very scary because using synchronous locks in asynchronous code can easily lead to a deadlock. However, I understand that it was easier to use compared to `RefCell` because `RefCell` is not Sync/Send and it would cause lots of problems and the need to use unsafe. Please add a comment which explains what is the purpose of those locks.

kfernandez31 replied:
I believe this comment:
```
/// One can only delegate a "mutating map" on an instance if all other "mutating maps" on it have ceased.
/// In other words, you must ensure sequential acquisition of mutable borrows to an instance.
/// Failing to comply with this rule will lead to a `panic!`, much in the same way that breaking the rule for
/// a `RefCell` would do.
```
kind of explains it, but I'll add another one over the locks just in case.

kfernandez31 commented:
v3: further comment resolution
@piodul review ping
